### PR TITLE
Add method to get the Version when a key was last written to

### PR DESCRIPTION
**Pull request by eddyashton**
See #2134 for discussion.

This implements option 2, with some unit tests for basic correctness and contention (checking we produce a monotonic chain after concurrent execution).

The name of this method is still up for grabs.

### CODE REVIEW

File: src/kv/test/kv_test.cpp [hunk 1]
```diff
@@ -149,6 +161,126 @@ TEST_CASE("Reads/writes and deletions")
   }
 }
 
+TEST_CASE("get_previous_version")
```
Comment (jumaffre) [inline src/kv/test/kv_test.cpp:164]: Can we also check that this is true across a primary and backup store and after deserialising snapshots?

File: src/kv/map_handle.h [hunk 2]
```diff
@@ -88,6 +88,26 @@ namespace kv
       return read_handle.has(KSerialiser::to_serialised(key));
     }
 
+    /** Get version when this key was last written to, by a previous
```
Comment (jumaffre) [inline src/kv/map_handle.h:91]: I think `kv/api.rst` will need to include `get_previous_version` as a member of `ReadableMapHandle` to show up in the docs.

File: src/kv/test/kv_test.cpp (hunk 1 above)
Comment (eddyashton) [reply src/kv/test/kv_test.cpp:164]: Modified the basic tests in `kv_serialisation.cpp` and `kv_snapshot.cpp` to test this - let me know if you think of any other awkward cases we should be testing.

File: src/kv/test/kv_snapshot.cpp [hunk 3]
```diff
@@ -79,21 +95,34 @@ TEST_CASE("Simple snapshot" * doctest::test_suite("snapshot"))
...
-    REQUIRE_EQ(new_store.current_version(), 2);
+    REQUIRE_EQ(new_store.current_version(), second_snapshot_version);
 
     auto tx1 = new_store.create_tx();
-    auto handle = tx1.rw<MapTypes::StringString>("public:string_map");
 
-    auto v = handle->get("foo");
-    REQUIRE(v.has_value());
-    REQUIRE_EQ(v.value(), "bar");
+    {
```
Comment (jumaffre) [inline src/kv/test/kv_snapshot.cpp:102]: Should we also test for key deletion in the second snapshot?

**Review by jumaffre — APPROVED** [review]

**Comment by achamayou** [comment]
@eddyashton can we have a changelog entry for this?

File: doc/build_apps/kv/api.rst [hunk 4]
```diff
@@ -43,7 +43,7 @@ Handle
 
 .. doxygenclass:: kv::ReadableMapHandle
    :project: CCF
-   :members: get, has, foreach
+   :members: get, has, foreach, get_version_of_previous_write
```
Comment (achamayou) [inline doc/build_apps/kv/api.rst:46]: fwiw I do like that name.

File: src/kv/map_handle.h [hunk 5]
```diff
@@ -88,6 +88,26 @@ namespace kv
...
+     * Returns nullopt when there is no value, because the key has no value
+     * (never existed or has been removed). Note that this is always talking
+     * about the version of previously committed state and not the same values
+     * as @c get or @c has - this transaction's pending writes have no version
+     * yet, and this method does not talk about them.
+     *
+     * @param key Key to read
+     *
+     * @return Optional containing version of applied transaction which last
+     * wrote at this key, or nullopt of this key has no associated value
```
Comment (achamayou) [inline src/kv/map_handle.h:103]:
```suggestion
     * wrote at this key, or nullopt if this key has no associated value
```

**Review by achamayou — APPROVED** [review]

**Pull request merged**